tiledb/sm/tile/tile_metadata_generator.cc

```diff
@@ -0,0 +1,571 @@
+      return false;
+
+    default:
+      return true;
+  }
+}
+
+bool TileMetadataGenerator::has_sum_metadata(
+    const Datatype type, const bool var_size, const uint64_t cell_val_num) {
+  // No sum for var sized attributes or cells with more than one value.
+  if (var_size || cell_val_num != 1)
+    return false;
+
+  // No sum for any, blob, or non ascii strings.
+  switch (type) {
+    case Datatype::ANY:
+    case Datatype::BLOB:
+    case Datatype::STRING_UTF8:
+    case Datatype::STRING_UTF16:
+    case Datatype::STRING_UTF32:
+    case Datatype::STRING_UCS2:
+    case Datatype::STRING_UCS4:
+    case Datatype::STRING_ASCII:
+      return false;
+
+    default:
+      return true;
+  }
+}
+
+template <class T>
+const std::tuple<void*, void*> TileMetadataGenerator::min_max(
+    const Tile* tile, const uint64_t cell_size) {
+  assert(tile != nullptr);
+
+  // Initialize defaults.
+  void* min = (void*)&metadata_generator_type_data<T>::min;
+  void* max = (void*)&metadata_generator_type_data<T>::max;
+
+  // Get pointer to the data and cell num.
+  auto values = static_cast<T*>(tile->buffer()->data());
+  auto cell_num = tile->buffer()->size() / cell_size;
+
+  // Process cell by cell.
+  for (uint64_t c = 0; c < cell_num; c++) {
+    min = *static_cast<T*>(min) < values[c] ? min : &values[c];
+    max = *static_cast<T*>(max) > values[c] ? max : &values[c];
+  }
+
+  return {min, max};
+}
+
+template <>
+const std::tuple<void*, void*> TileMetadataGenerator::min_max<char>(
+    const Tile* tile, const uint64_t cell_size) {
+  assert(tile != nullptr);
+
+  // For strings, return null for empty tiles.
+  auto size = tile->buffer()->size();
+  if (size == 0)
+    return {nullptr, nullptr};
+
+  // Get pointer to the data, set the min max to the first value.
+  auto data = (char*)tile->buffer()->data();
+  void* min = data;
+  void* max = data;
+
+  // Process all cells, starting at the second value.
+  auto value = data + cell_size;
+  auto cell_num = size / cell_size;
+  for (uint64_t c = 1; c < cell_num; c++) {
+    min = strncmp((const char*)min, (const char*)value, size) > 0 ? value : min;
+    max = strncmp((const char*)max, (const char*)value, size) < 0 ? value : max;
+    value += cell_size;
+  }
+
+  return {min, max};
+}
+
+template <class T>
+const std::tuple<void*, void*, uint64_t>
+TileMetadataGenerator::min_max_nullable(
+    const Tile* tile, const Tile* tile_validity, const uint64_t cell_size) {
+  assert(tile != nullptr);
+
+  // Initialize defaults.
+  void* min = (void*)&metadata_generator_type_data<T>::min;
+  void* max = (void*)&metadata_generator_type_data<T>::max;
+  uint64_t null_count = 0;
+
+  // Get pointers to the data and cell num.
+  auto values = static_cast<T*>(tile->buffer()->data());
+  auto validity_values = static_cast<uint8_t*>(tile_validity->buffer()->data());
+  auto cell_num = tile->buffer()->size() / cell_size;
+
+  // Process cell by cell.
+  for (uint64_t c = 0; c < cell_num; c++) {
```

Comment:
Is the idea that we will always parallelize across tiles or attributes, and never within a tile? Because ideally we should follow the same principles around distributing threads based on the number of attributes, number of tiles and tile capacity.

Reply:
@stavrospapadopoulos yes, I'm following the same pattern the writer has right now, which is parallelizing on tiles or attributes.

Reply:
OK, thanks, just making a note that eventually we may have to parallelize over cells too.